### DISCUSSION TITLE
q-table does not display data when updating data

### DISCUSSION THREAD

**Discussion by dcapeletti** [quasarframework/quasar#16289] [General - Components / Directives / etc]
Hi, I have a problem with q-table. When the component is loaded with page=1 the data is displayed correctly, but when the page is changed to get the next data, when getting correctly the same data is not displayed in the table.
When the component is loaded the data is displayed correctly in the table.
![page1](https://github.com/quasarframework/quasar/assets/5890886/dcf90ca7-05fb-4d84-8623-119acaaca5f3)

But when you go to page 2 with the paging control, the data is not rendered in the table even though the data has been downloaded and installed in the pinia store. "storeProductos.listaProductos" is the array in pinia that holds the list of objects returned by the server.
![page2](https://github.com/quasarframework/quasar/assets/5890886/5a45de59-1ac6-4619-a2c5-f4c6df2509ab)
![page2-1](https://github.com/quasarframework/quasar/assets/5890886/3382cd4f-d060-47c5-ae6c-1ffbd8f2ace3)
![page2-2](https://github.com/quasarframework/quasar/assets/5890886/d05cf421-5839-48f0-963c-35f6cb00471d)

My code:
```
<template>
  <div class="text-h6 q-mb-md">
    <q-table
      table-header-class="text-bold bg-grey"
      class="my-sticky-header-table"
      flat
      dense
      separator="cell"
      bordered
      :rows="storeProductos.listaProductos"
      :columns="columns"
      row-key="idProducto"
      v-model:pagination="pagination"
      :visible-columns="columnasVisibles"
      :loading="cargando"
      no-data-label="No hay datos que mostrar..."
      loading-label="Cargando datos..."
      rows-per-page-label="Registros a mostrar:"
    >
      <div class="row justify-center q-mt-md">
        <q-pagination
          v-model="pagination.page"
          color="grey-8"
          :max="numberPages"
          size="sm"
        />
      </div>

      <template v-slot:body="props">
        <q-tr :props="props">
          <q-td key="Nombre" :props="props">
            {{ props.row.Nombre }}
            <q-popup-edit
              v-model="props.row.Nombre"
              v-slot="scope"
              @save="
                (val, initialValue) =>
                  guardar(val, initialValue, props, 'Nombre')
              "
            >
              <q-input
                v-model="scope.value"
                dense
                autofocus
                counter
                @keyup.enter="scope.set"
              />
            </q-popup-edit>
          </q-td>

          <q-td key="Marca" :props="props">
            {{ props.row.Marca }}
            <q-popup-edit
              v-model="props.row.Marca"
              v-slot="scope"
              @save="
                (val, initialValue) =>
                  guardar(val, initialValue, props, 'Marca')
              "
            >
              <q-input
                v-model="scope.value"
                dense
                autofocus
                counter
                @keyup.enter="scope.set"
              />
            </q-popup-edit>
          </q-td>

          <q-td key="Peso_Kg" :props="props">
            {{ props.row.Peso_Kg }}
            <q-popup-edit
              v-model.number="props.row.Peso_Kg"
              v-slot="scope"
              :validate="(val) => !isNaN(val) && val != ''"
              @hide="(val) => !isNaN(val) && val != ''"
              @save="
                (val, initialValue) =>
                  guardar(val, initialValue, props, 'Peso_Kg')
              "
            >
              <q-input
                v-model="scope.value"
                dense
                type="number"
                :error="!isNaN(scope.value) && scope.value == ''"
                :error-message="
                  !isNaN(scope.value) && scope.value == ''
                    ? 'Solo ingrese números'
                    : ''
                "
                autofocus
                counter
                @keyup.enter="scope.set"
              />
            </q-popup-edit>
          </q-td>

          <q-td key="Imagen" :props="props">
            <q-btn
              rounded
              color="primary"
              icon="image"
              @click="
                mostrarImagen(
                  props.row.idProducto,
                  props.row.Nombre,
                  props.row.Imagen
                )
              "
            />
          </q-td>

          <q-td key="Precio_U" :props="props">
            <div v-if="props.row.Precio_U != null">
              {{ "$" + props.row.Precio_U }}
            </div>
            <q-popup-edit
              v-model="props.row.Precio_U"
              v-slot="scope"
              :validate="(val) => !isNaN(val) && val != ''"
              @hide="(val) => !isNaN(val) && val != ''"
              @save="
                (val, initialValue) =>
                  guardar(val, initialValue, props, 'Precio_U')
              "
            >
              <q-input
                v-model="scope.value"
                dense
                autofocus
                type="number"
                :error="!isNaN(scope.value) && scope.value == ''"
                :error-message="
                  !isNaN(scope.value) && scope.value == ''
                    ? 'Solo ingrese números'
                    : ''
                "
                counter
                @keyup.enter="scope.set"
              />
            </q-popup-edit>
          </q-td>

          <q-td key="Comision" :props="props">
            <div v-if="props.row.Comision != null">
              {{ props.row.Comision + "%" }}
            </div>
            <q-popup-edit
              v-model.number="props.row.Comision"
              v-slot="scope"
              :validate="(val) => !isNaN(val) && val != ''"
              @hide="(val) => !isNaN(val) && val != ''"
              @save="
                (val, initialValue) =>
                  guardar(val, initialValue, props, 'Comision')
              "
            >
              <q-input
                v-model.number="scope.value"
                dense
                type="number"
                :error="!isNaN(scope.value) && scope.value == ''"
                :error-message="
                  !isNaN(scope.value) && scope.value == ''
                    ? 'Solo ingrese números'
                    : ''
                "
                autofocus
                counter
                @keyup.enter="scope.set"
              />
            </q-popup-edit>
          </q-td>

          <q-td key="IVA" :props="props">
            {{ props.row.IVA }}
            <q-popup-edit
              v-model="props.row.IVA"
              v-slot="scope"
              :validate="(val) => !isNaN(val) && val != ''"
              @hide="(val) => !isNaN(val) && val != ''"
              @save="
                (val, initialValue) => guardar(val, initialValue, props, 'IVA')
              "
            >
              <q-input
                v-model="scope.value"
                dense
                autofocus
                counter
                type="number"
                :error="!isNaN(scope.value) && scope.value == ''"
                :error-message="
                  !isNaN(scope.value) && scope.value == ''
                    ? 'Solo ingrese números'
                    : ''
                "
                @keyup.enter="scope.set"
              />
            </q-popup-edit>
          </q-td>

          <q-td key="Tipo_Envase_Retornable" :props="props">
            <div class="q-gutter-md">
              <q-select
                v-model="props.row.Tipo_Envase_Retornable_id"
                dense
                use-chips
                :options="storeTipoEnvases.listaTipoEnvases"
                option-label="Nombre"
                option-value="idTipo_Envase"
                @remove="
                  guardar(
                    props.row.Tipo_Envase_Retornable,
                    initialValue,
                    props,
                    'Tipo_Envase_Retornable'
                  )
                "
                @popup-hide="
                  guardar(
                    props.row.Tipo_Envase_Retornable,
                    initialValue,
                    props,
                    'Tipo_Envase_Retornable'
                  )
                "
              />
            </div>
          </q-td>

          <q-td key="Precio_U_Manual" :props="props">
            <div class="q-pa-md">
              <q-checkbox
                dense
                v-model="props.row.Precio_U_Manual"
                @click="
                  guardar(
                    props.row.Precio_U_Manual,
                    initialValue,
                    props,
                    'Precio_U_Manual'
                  )
                "
              />
            </div>
          </q-td>

          <q-td key="Precio_U_Por_Costos_Mas_Comision" :props="props">
            <div class="q-pa-md">
              <q-checkbox
                dense
                v-model="props.row.Precio_U_Por_Costos_Mas_Comision"
                @click="
                  guardar(
                    props.row.Precio_U_Por_Costos_Mas_Comision,
                    initialValue,
                    props,
                    'Precio_U_Por_Costos_Mas_Comision'
                  )
                "
              >
                <q-tooltip
                  >Active esta opción si desea que el precio sea recalculado
                  automaticamente.</q-tooltip
                >
              </q-checkbox>
            </div>
          </q-td>

          <q-td key="Precio_U_Por_Categoria" :props="props">
            <div class="q-pa-md">
              <q-checkbox
                dense
                v-model="props.row.Precio_U_Por_Categoria"
                @click="
                  guardar(
                    props.row.Precio_U_Por_Categoria,
                    initialValue,
                    props,
                    'Precio_U_Por_Categoria'
                  )
                "
              />
            </div>
          </q-td>
          <q-td key="Delete" :props="props">
            <q-btn
              rounded
              color="red"
              icon="delete"
              @click="eliminarProducto(props.row.idProducto, props.row.Nombre)"
            />
          </q-td>
        </q-tr>
      </template>

      <template v-slot:top="props">
        <div class="col-15 q-table__title">Productos</div>

        <q-space />

        <q-select
          v-model="columnasVisibles"
          multiple
          outlined
          dense
          options-dense
          display-value="Columnas"
          emit-value
          map-options
          :options="columns"
          option-value="name"
          options-cover
          style="min-width: 150px"
        />

        <q-btn
          flat
          round
          dense
          icon="refresh"
          @click="listarProductos()"
          class="q-mr-mr"
        />

        <q-btn
          flat
          round
          dense
          :icon="props.inFullscreen ? 'fullscreen_exit' : 'fullscreen'"
          @click="props.toggleFullscreen"
          class="q-mr-mr"
        />
      </template>

      <template v-slot:pagination>
        <q-btn
          v-if="numberPages > 1"
          icon="first_page"
          color="grey-8"
          round
          dense
          flat
          :disable="pagination.page <= 0"
          @click="firstPage"
        />

        <q-btn
          icon="chevron_left"
          color="grey-8"
          round
          dense
          flat
          :disable="pagination.page <= 1"
          @click="prevPage"
        />

        <q-btn
          icon="chevron_right"
          color="grey-8"
          round
          dense
          flat
          :disable="pagination.page >= numberPages"
          @click="nextPage"
        />

        <q-btn
          v-if="numberPages > 1"
          icon="last_page"
          color="grey-8"
          round
          dense
          flat
          :disable="pagination.page >= numberPages"
          @click="lastPage"
        />
      </template>
    </q-table>
    <!-- Total Productos {{ storeProductos.totalProductos }} Total Pagina
    {{ numberPages }} En Pagina {{ pagination.page }} -->
  </div>
</template>

<script>
import { defineComponent, ref, computed, onMounted, watch } from "vue";
import useApiProductos from "../../api/apiProductos";
import useApiTipoEnvases from "../../api/apiTipoEnvases";
import { useStoreProductos } from "../../stores/storeProductos";
import { useStoreTipoEnvases } from "../../stores/storeTipoEnvase";
import { Notify, useQuasar } from "quasar";
import dialogEliminarProducto from "./dialogEliminarProducto.vue";
import dialogImagen from "./dialogImagen.vue";

export default defineComponent({
  name: "tblListaProductos",

  setup() {
    const $q = useQuasar();
    const storeProductos = useStoreProductos();
    const apiProductos = useApiProductos();
    const apiTipoEnvases = useApiTipoEnvases();
    const storeTipoEnvases = useStoreTipoEnvases();
    const columns = [
      {
        name: "Nombre",
        required: true,
        label: "Nombre",
        align: "left",
        field: (row) => row.Nombre,
        format: (val) => `${val}`,
        sortable: true,
      },
      {
        name: "Marca",
        required: true,
        label: "Marca",
        align: "left",
        field: "Marca",
        format: (val) => `${val}`,
        sortable: true,
      },
      {
        name: "Peso_Kg",
        required: false,
        label: "Peso_Kg",
        align: "left",
        field: "DnPeso_Kgi",
        format: (val) => `${val}`,
        sortable: false,
      },
      {
        name: "Imagen",
        align: "left",
        label: "Imagen",
        field: "Imagen",
        format: (val) => `${val}`,
        sortable: true,
      },
      {
        name: "Precio_U",
        align: "left",
        label: "Precio U",
        field: "Precio_U",
        format: (val) => `${val}`,
        sortable: true,
      },
      {
        name: "Comision",
        align: "left",
        label: "Comision",
        field: "Comision",
        format: (val) => `${val}'`,
        sortable: true,
      },
      {
        name: "IVA",
        align: "left",
        label: "IVA",
        field: "IVA",
        format: (val) => `${val + "%"}`,
        sortable: true,
      },
      {
        name: "Tipo_Envase_Retornable",
        align: "left",
        label: "Tipo Envase",
        field: "Tipo_Envase_Retornable",
        format: (val) => `${val}`,
        sortable: false,
      },
      {
        name: "Precio_U_Manual",
        align: "left",
        label: "Precio Manual",
        field: "Precio_U_Manual",
        format: (val) => `${val}`,
        sortable: true,
      },
      {
        name: "Precio_U_Por_Costos_Mas_Comision",
        align: "left",
        label: "Precio calculado",
        field: "Precio_U_Por_Costos_Mas_Comision",
        format: (val) => `${val}`,
        sortable: true,
      },
      {
        name: "Precio_U_Por_Categoria",
        align: "left",
        label: "Precio por Cat.",
        field: "Precio_U_Por_Categoria",
        format: (val) => `${val}`,
        sortable: true,
      },
      {
        name: "Delete",
        label: "Eliminar",
        align: "left",
      },
    ];

    const columnasVisibles = ref([
      "Nombre",
      "Marca",
      "Peso_Kg",
      "Imagen",
      "Precio_U",
      "Comision",
      "IVA",
      "Tipo_Envase_Retornable",
      "Precio_U_Manual",
      "Precio_U_Por_Costos_Mas_Comision",
      "Precio_U_Por_Categoria",
      "Delete",
    ]);

    const pagination = ref({
      sortBy: "asc",
      descending: false,
      page: 1,
      rowsPerPage: 5,
    });

    const cargando = ref(false);

    const numberPages = computed(() => {
      return Math.ceil(
        storeProductos.totalProductos / pagination.value.rowsPerPage
      );
    });

    function guardar(newVal, oldVal, fila, col) {
      //Tengo que buscar que columna se edito...
      const productoEditado = storeProductos.listaProductos.find(
        (obj) => obj.idProducto == fila.row.idProducto
      );
      console.log(productoEditado);
      switch (col) {
        case "Nombre":
          productoEditado.Nombre = newVal;
          break;
        case "Marca":
          productoEditado.Marca = newVal;
          break;
        case "Peso_Kg":
          productoEditado.Peso_Kg = newVal;
          break;
        case "Imagen":
          productoEditado.Imagen = newVal;
          break;
        case "Precio_U":
          productoEditado.Precio_U = newVal;
          break;
        case "Comision":
          productoEditado.Comision = newVal;
          break;
        case "IVA":
          productoEditado.IVA = newVal;
          break;
      }

      apiProductos
        .editarProducto(productoEditado.idProducto, productoEditado)
        .then((response) => {
          console.log(response.data);
          Notify.create({
            message: "Producto editado con éxito...",
            icon: "check",
            color: "green",
          });
          // storeProductos.totalProductos = response.data.totalProductos
        })
        .catch((error) => {
          Notify.create({
            message: "Error al editar el producto:" + error,
            color: "red",
            icon: "error",
          });
          apiProductos.getProductos(
            pagination.value.page,
            pagination.value.rowsPerPage
          );
        });
    }

    function listarProductos() {
      cargando.value = true;
      apiProductos
        .getProductos(pagination.value.page, pagination.value.rowsPerPage)
        .then((response) => {
          cargando.value = false;
        });
    }

    function eliminarProducto(idProducto, NombreProd) {
      $q.dialog({
        component: dialogEliminarProducto,
        componentProps: { modelvalue: idProducto, Nombre: NombreProd },
      });
      // console.log("eliminando cliente ", idProducto);
    }

    function mostrarImagen(idProducto, NombreProd, img) {
      $q.dialog({
        component: dialogImagen,
        componentProps: {
          modelvalue: idProducto,
          producto: NombreProd,
          imagen: img,
        },
      });
      // console.log("eliminando cliente ", idProducto);
    }

    onMounted(() => {
      // pagination.value.page = 0;
      listarProductos();
      apiTipoEnvases.getTipoEnvases(1, 100);
      // cargando.value = false;
    });

    function firstPage() {
      pagination.value.page = 1;
      // console.log(pagination.value);
      listarProductos();
    }

    function prevPage() {
      pagination.value.page = pagination.value.page - 1;
      // console.log(pagination.value);
      listarProductos();
    }

    function nextPage() {
      pagination.value.page = pagination.value.page + 1;
      // console.log(pagination.value);
      listarProductos();
    }

    function lastPage() {
      pagination.value.page = numberPages.value;
      // console.log(pagination.value);
      listarProductos();
    }

    return {
      columns,
      pagination,
      numberPages,
      firstPage,
      prevPage,
      nextPage,
      lastPage,
      storeProductos,
      storeTipoEnvases,
      guardar,
      eliminarProducto,
      mostrarImagen,
      columnasVisibles,
      cargando,
      listarProductos,
    };
  },
});
</script>

```

However, if I add rowsNumber to the pagination, the data is displayed correctly and everything works fine, but I cannot sort it when I click on the column.

```
    const pagination = ref({
      sortBy: "desc",
      descending: false,
      page: 1,
      rowsPerPage: 5,
      rowsNumber: null,
    });
```

I have been looking for a long time but I have not found the solution, I have no idea why q-table does not load the data when you change the data (go to another page and download new data). I need your help, thanks for your time. Thanks

@quasar/cli v2.2.3
Pkg quasar................ v2.12.2
Pkg @quasar/app-webpack... v3.9.2

**Comment by metalsadman** [comment]
tldr, try to reproduce it in a codepen, I'm fairly certain this is not a qtable issue tho. one thing that that i noticed in your code is your use of callbacks, whereas I would use async awaits instead, much easier to follow.
